server/Dockerfile

```diff
@@ -0,0 +1,60 @@
+# create base image with all packages, proper entrypoint, and directories created
+FROM node:erbium-alpine AS base
+
+# install any packages we need from apt here
+RUN apk add --update dumb-init
+
+# set entrypoint to `dumb-init` as it handles being pid 1 and forwarding signals
+# so that you dont need to bake that logic into your node app
+ENTRYPOINT ["dumb-init", "--"]
+
+# all of our code will live in `/app`
+WORKDIR /app
+
+# using the base image, create an image containing all of our files
+# and dependencies installed, devDeps and test directory included
+FROM base AS dependencies
+
+COPY package*.json ./
+RUN npm set progress=false \
+    && npm config set depth 0 \
+    && npm i
+
+COPY .eslintrc.json .
+COPY ./server ./server
+COPY ./test ./test
+COPY ./lib ./lib
+
+# `node` user is created in base node image
+# we want to use this non-root user for running the server in case of attack
+# https://github.com/nodejs/docker-node/blob/master/docs/BestPractices.md#non-root-user
+USER node
+
+# if you have any build scripts to run, like for the `templated-site` flavor
+# uncomment and possibly modify the following RUN command:
+# RUN npm run build
+# keeping all of the bash commands you can within a single RUN is generally important,
+# but for this case it's likely that we want to use the cache from the prune which will
+# change infrequently.
+
+# test running image using all of the files and devDeps
+FROM dependencies AS test
+ENV NODE_ENV=test
```

Comment:
What similarly provides the release stage `NODE_ENV=production`?

Reply:
The orchestration framework you use should provide that. I people use that differently, so I don't want to define it in the Dockerfile except for in the test stage.

For example, where I work uses `production` for all non-development envs. I personally prefer `development`, `staging`, and `production`.

Reply:
The boilerplate originally referenced `production` and `development`, but I moved away from that upon closer reading of the 12 factor approach to configuration and environments (https://12factor.net/config).  I consider `production` and `test` reasonable exceptions just due to conventions in nodejs.

As for setting `NODE_ENV` to `production` for release, I am actually still interested in doing so, partly because the application may not run without the pruned dependencies unless `NODE_ENV` is `production` (which does happen to be the case with hpal-debug).  The dockerfile's ENVs also seem to be the "least specific", so all other methods including orchestration should be able to override them anyway.

Reply:
Good call re: pruned deps. Adding that node env now.